File: scripts/generate-course-draft.ts
/**
 * Generates DRAFT editorial suggestions (overview / outcomes / curriculum
 * presentation) for every real catalogue course, from nothing but course
 * name, category, subcategory, duration and the existing verified
 * curriculum/topics — never from guessed facts.
 *
 * This is a proposal file only:
 *   - It does NOT write to `src/content/catalogue/enrichment.ts`.
 *   - It does NOT change anything the website renders.
 *   - Every row is default-tagged `draftStatus: "NEEDS REVIEW"` — nothing
 *     here is meant to be imported until a human (or a later, explicit
 *     approval step) marks it `APPROVED`.
 *
 * Fields the task explicitly says never to draft (mode, level, eligibility,
 * certification, fees, accreditation, recognition, admission requirements,
 * placement/income claims) are left exactly as they already are in the live
 * data — populated only where already verified, blank otherwise. This
 * script never writes to those columns.
 *
 * Run with: pnpm dlx tsx scripts/generate-course-draft.ts
 */
import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { catalogueCourses } from "@/content/catalogue/courses";
import { masterCourseTemplate, type MasterCourseRow } from "@/content/catalogue/master-template";
import type { CourseCurriculumModule } from "@/content/catalogue/types";

const OUT_DIR = join(process.cwd(), "content-templates");
mkdirSync(OUT_DIR, { recursive: true });

/** Course-type words that must never be treated as proof of accreditation, recognition, or licensing. */
const SENSITIVE_COURSE_TYPES = new Set(["Certificate", "Diploma", "Advanced Diploma", "Training Program"]);

const OUTCOME_VERBS = ["Understand", "Learn to", "Practice", "Explore", "Use basic", "Develop familiarity with"];

type DraftRow = MasterCourseRow & {
  overviewDraft: string | null;
  outcomesDraft: string[] | null;
  curriculumDraft: CourseCurriculumModule[] | null;
  draftStatus: "NEEDS REVIEW" | "APPROVED" | "REJECTED";
  draftFlags: string[];
};

/**
 * `splitCurriculum` (helpers.ts) splits the spreadsheet's raw prose on every
 * comma, which breaks apart anything like "MS Office (Word, Excel,
 * PowerPoint)" into three fragments with an unclosed paren. Re-joins those
 * fragments back into one topic before this script uses them for drafting —
 * purely a formatting repair, no topic content added or removed.
 */
function repairSplitParens(topics: string[]): string[] {
  const out: string[] = [];
  let buffer: string | null = null;
  for (const t of topics) {
    if (buffer !== null) {
      buffer = `${buffer}, ${t}`;
      if (t.includes(")")) {
        out.push(buffer);
        buffer = null;
      }
      continue;
    }
    if (t.includes("(") && !t.includes(")")) {
      buffer = t;
    } else {
      out.push(t);
    }
  }
  if (buffer !== null) out.push(buffer); // unclosed paren at end of list — keep as-is rather than dropping content
  return out;
}

/** Trims a topic string down to a short noun phrase suitable for inline listing, without altering its meaning. */
function shortenTopic(topic: string): string {
  return topic
    .replace(/\s*\([^)]*\)\s*/g, " ") // drop parenthetical examples — still fine to mention the parent topic
    .replace(/\s+/g, " ")
    .trim();
}

function naturalJoin(items: string[]): string {
  if (items.length === 0) return "";
  if (items.length === 1) return items[0];
  if (items.length === 2) return `${items[0]} and ${items[1]}`;
  return `${items.slice(0, -1).join(", ")} and ${items[items.length - 1]}`;
}

/** Picks a small, representative sample of topics for prose — first few, deduped, shortened. */
function sampleTopics(topics: string[], max: number): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const raw of topics) {
    const t = shortenTopic(raw);
    const key = t.toLowerCase();
    if (!t || seen.has(key)) continue;
    seen.add(key);
    out.push(t);
    if (out.length >= max) break;
  }
  return out;
}

function areaLabel(row: MasterCourseRow): string {
  return row.subcategory ?? row.category ?? "this subject area";
}

/**
 * Builds a factual, neutral overview strictly from curriculum topics — no
 * job-readiness, guarantee, certification, or recognition language, and no
 * course-type word treated as proof of anything.
 */
function draftOverview(row: MasterCourseRow): { text: string | null; flag: string | null } {
  if (row.overview) return { text: null, flag: null }; // already verified — do not touch or duplicate
  if (row.curriculum.length === 0) return { text: null, flag: "No curriculum topics available — insufficient source data to draft an overview." };

  const sample = sampleTopics(repairSplitParens(row.curriculum), 6);
  if (sample.length === 0) return { text: null, flag: "Curriculum topics too sparse to draft an overview." };

  const topicPhrase = naturalJoin(sample);
  const text = `This course introduces learners to ${areaLabel(row).toLowerCase()}, covering ${topicPhrase}, over a duration of ${row.duration}.`;

  return { text, flag: null };
}

/**
 * Generates 3–6 outcomes, one per distinct topic sampled from the
 * curriculum, cycling through approved neutral verbs. Never uses "Master",
 * "expert", "guaranteed", or "job-ready" wording.
 */
function draftOutcomes(row: MasterCourseRow): { list: string[] | null; flag: string | null } {
  if (row.outcomes && row.outcomes.length > 0) return { list: null, flag: null }; // already verified
  if (row.curriculum.length === 0) return { list: null, flag: "No curriculum topics available — insufficient source data to draft outcomes." };

  const repaired = repairSplitParens(row.curriculum);
  const sample = sampleTopics(repaired, 6).slice(0, Math.max(3, Math.min(6, repaired.length)));
  if (sample.length < 3) {
    return {
      list: null,
      flag: `Only ${sample.length} distinct topic(s) in the curriculum — not enough to safely draft 3–6 outcomes.`,
    };
  }

  const list = sample.map((topic, i) => `${OUTCOME_VERBS[i % OUTCOME_VERBS.length]} ${topic.toLowerCase()}.`);
  return { list, flag: null };
}

/**
 * Only regroups the flat topic list into modules when a boundary is
 * literally present in the source text (a topic ending in "Introduction"
 * marking the start of a new named subject) — never an invented split.
 */
function draftCurriculum(
  slug: string,
  row: MasterCourseRow,
): { modules: CourseCurriculumModule[] | null; flag: string | null } {
  const course = catalogueCourses.find((c) => c.slug === slug);
  const alreadyStructured = course ? course.curriculumModules !== null : slug === "full-stack-web-development";
  if (alreadyStructured) {
    return { modules: null, flag: null }; // already has a richer structure — nothing to draft
  }

  const topics = repairSplitParens(row.curriculum);
  if (topics.length === 0) return { modules: null, flag: "No curriculum topics available." };

  const boundaries = topics
    .map((t, i) => ({ i, isBoundary: i > 0 && /\bIntroduction$/i.test(t.trim()) }))
    .filter((x) => x.isBoundary)
    .map((x) => x.i);

  if (boundaries.length === 0) {
    return { modules: null, flag: "No obvious grouping present in the source topic list — kept as a flat list." };
  }

  const starts = [0, ...boundaries];
  const modules: CourseCurriculumModule[] = starts.map((start, idx) => {
    const end = idx + 1 < starts.length ? starts[idx + 1] : topics.length;
    const chunk = topics.slice(start, end);
    const marker = chunk[0];
    const title = /\bIntroduction$/i.test(chunk[0].trim())
      ? `${chunk[0].trim().replace(/\s*Introduction$/i, "")} Concepts`
      : idx === 0
        ? "Foundations"
        : `${marker} Concepts`;
    return { title, topics: chunk };
  });

  return { modules, flag: null };
}

function isSensitiveType(courseType: string | null): boolean {
  return courseType !== null && SENSITIVE_COURSE_TYPES.has(courseType);
}

const draftRows: DraftRow[] = masterCourseTemplate.map((row) => {
  const flags: string[] = [];

  const overview = draftOverview(row);
  if (overview.flag) flags.push(overview.flag);

  const outcomes = draftOutcomes(row);
  if (outcomes.flag) flags.push(outcomes.flag);

  const curriculum = draftCurriculum(row.slug, row);
  if (curriculum.flag) flags.push(curriculum.flag);

  if (isSensitiveType(row.courseType)) {
    flags.push(
      `Course type "${row.courseType}" — title alone is not treated as proof of accreditation, government recognition, university affiliation, internship placement, or professional licensing; overview/outcomes kept neutral.`,
    );
  }

  return {
    ...row,
    overviewDraft: overview.text,
    outcomesDraft: outcomes.list,
    curriculumDraft: curriculum.modules,
    draftStatus: "NEEDS REVIEW",
    draftFlags: flags,
  };
});

// --- Write JSON intermediate for the Python XLSX step ---
writeFileSync(join(OUT_DIR, "course-content-draft.json"), JSON.stringify(draftRows, null, 2), "utf-8");

// --- Console summary for the calling task ---
const totalCourses = draftRows.length;
const overviewsGenerated = draftRows.filter((r) => r.overviewDraft !== null).length;
const outcomesGenerated = draftRows.filter((r) => r.outcomesDraft !== null).length;
const curriculumDrafted = draftRows.filter((r) => r.curriculumDraft !== null).length;
const insufficientData = draftRows.filter(
  (r) => r.overviewDraft === null && r.outcomesDraft === null && r.curriculumDraft === null && r.overview === null,
).length;
const flaggedCourses = draftRows.filter((r) => r.draftFlags.length > 0).length;

console.log(
  JSON.stringify(
    {
      totalCourses,
      overviewsGenerated,
      outcomesGenerated,
      curriculumDrafted,
      insufficientData,
      flaggedCourses,
    },
    null,
    2,
  ),
);

File: scripts/generate-course-draft.py
"""
Builds content-templates/course-content-draft.xlsx from the JSON written by
generate-course-draft.ts. Run the TS script first (it reads the live
catalogue and drafts overview/outcomes/curriculum suggestions); this step
only formats what it wrote — it never invents or edits draft content itself.

    pnpm dlx tsx scripts/generate-course-draft.ts
    python3 scripts/generate-course-draft.py

Every row lands with draftStatus = "NEEDS REVIEW". Nothing in this workbook
is read by the website or by the import workflow (scripts/import-course-content.py
only reads columns matching its known template shape) until someone marks a
row APPROVED and its content is manually carried into
content-templates/course-content-template.xlsx (or a future draft-import
step is explicitly requested).
"""
import json
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
from openpyxl.utils import get_column_letter
from openpyxl.worksheet.datavalidation import DataValidation

ROOT = Path(__file__).resolve().parent.parent
OUT_DIR = ROOT / "content-templates"

with open(OUT_DIR / "course-content-draft.json", encoding="utf-8") as f:
    rows = json.load(f)

# Verified/source columns (unchanged from the master template) + new draft columns.
COLUMNS = [
    "slug", "courseCode", "courseName", "category", "subcategory", "courseType",
    "duration", "curriculum",
    "mode", "level", "overview", "outcomes", "eligibility", "certification", "fees",
    "featured", "status", "notes",
    "overviewDraft", "outcomesDraft", "curriculumDraft", "draftStatus", "draftFlags",
]

DRAFT_COLUMNS = {"overviewDraft", "outcomesDraft", "curriculumDraft", "draftStatus", "draftFlags"}
VERIFIED_TRACKED = {"mode", "level", "overview", "outcomes", "eligibility", "certification", "fees"}


def flatten_curriculum_modules(modules):
    """[{title, topics}] -> 'Title >> topic | topic || Title2 >> topic' (matches the importer's richer-curriculum syntax)."""
    if not modules:
        return ""
    return " || ".join(f"{m['title']} >> " + " | ".join(m["topics"]) for m in modules)


def cell_value(row, key):
    if key == "curriculum":
        return " | ".join(row["curriculum"])
    if key == "curriculumDraft":
        return flatten_curriculum_modules(row.get("curriculumDraft"))
    if key == "outcomes" or key == "outcomesDraft" or key == "eligibility":
        value = row.get(key)
        return " | ".join(value) if value else ""
    if key == "draftFlags":
        return " | ".join(row.get("draftFlags", []))
    if key == "featured":
        return "TRUE" if row["featured"] else "FALSE"
    value = row.get(key)
    if value is None:
        return ""
    return str(value)


wb = Workbook()

# --- Sheet 1: Instructions --------------------------------------------------
instructions = wb.active
instructions.title = "Instructions"
instructions.sheet_view.showGridLines = False
instructions.column_dimensions["A"].width = 100

title_font = Font(name="Arial", size=14, bold=True)
body_font = Font(name="Arial", size=11)
bold_font = Font(name="Arial", size=11, bold=True)

lines = [
    ("SRS Academy — Course Content Draft Sheet (for review)", title_font),
    ("", body_font),
    ("What this is", bold_font),
    ("- Editorial DRAFT suggestions for overview, outcomes and curriculum presentation,", body_font),
    ("  generated only from course name, category, subcategory, duration and the", body_font),
    ("  existing verified curriculum topics. Nothing here has been published — this", body_font),
    ("  workbook does not change the live website and was not written to enrichment.ts.", body_font),
    ("", body_font),
    ("How to use it", bold_font),
    ("- Review each row's overviewDraft / outcomesDraft / curriculumDraft.", body_font),
    ("- Set draftStatus to APPROVED once you're satisfied a row is accurate, or", body_font),
    ("  REJECTED to discard it. Rows stay NEEDS REVIEW until you change them.", body_font),
    ("- Nothing marked NEEDS REVIEW may be imported into the live enrichment source.", body_font),
    ("- draftFlags calls out anything uncertain: too few curriculum topics to draft", body_font),
    ("  outcomes safely, no obvious grouping for curriculumDraft, or a course-type", body_font),
    ("  word (Certificate / Diploma / Advanced Diploma / Training Program) that must", body_font),
    ("  never be read as proof of accreditation, government recognition, university", body_font),
    ("  affiliation, internship placement, or professional licensing.", body_font),
    ("- Columns mode / level / eligibility / certification / fees are intentionally", body_font),
    ("  blank unless already verified — this pass never drafts or infers them.", body_font),
    ("- overview / outcomes stay blank in this sheet when a draft was generated for", body_font),
    ("  them (see overviewDraft / outcomesDraft instead); where overview/outcomes", body_font),
    ("  already show a value, that value is already verified and was left untouched —", body_font),
    ("  no draft was generated for that field.", body_font),
    ("", body_font),
    ("Getting an approved draft into the live site", bold_font),
    ("- Approving a row here does not publish it. Copy the approved overview/outcomes/", body_font),
    ("  curriculum text into content-templates/course-content-template.xlsx (or an", body_font),
    ("  edited copy of it) and run the existing two-step import workflow:", body_font),
    ("    python3 scripts/import-course-content.py <edited.xlsx>          (preview)", body_font),
    ("    python3 scripts/import-course-content.py <edited.xlsx> --apply  (publish)", body_font),
]
for i, (text, font) in enumerate(lines, start=1):
    cell = instructions.cell(row=i, column=1, value=text)
    cell.font = font
    cell.alignment = Alignment(wrap_text=False, vertical="top")

total = len(rows)
with_overview_draft = sum(1 for r in rows if r.get("overviewDraft"))
with_outcomes_draft = sum(1 for r in rows if r.get("outcomesDraft"))
with_curriculum_draft = sum(1 for r in rows if r.get("curriculumDraft"))
flagged = sum(1 for r in rows if r.get("draftFlags"))

row_after = len(lines) + 2
instructions.cell(row=row_after, column=1, value="Generation summary").font = bold_font
instructions.cell(
    row=row_after + 1,
    column=1,
    value=(
        f"{total} courses processed — overviewDraft: {with_overview_draft}, "
        f"outcomesDraft: {with_outcomes_draft}, curriculumDraft: {with_curriculum_draft}, "
        f"rows with a flag: {flagged}"
    ),
).font = body_font

# --- Sheet 2: Courses --------------------------------------------------------
ws = wb.create_sheet("Courses")
ws.sheet_view.showGridLines = False
ws.freeze_panes = "A2"

header_fill = PatternFill(start_color="1F2937", end_color="1F2937", fill_type="solid")
header_font = Font(name="Arial", size=10, bold=True, color="FFFFFF")
draft_header_fill = PatternFill(start_color="1E5A8E", end_color="1E5A8E", fill_type="solid")
verified_tracked_fill = PatternFill(start_color="92722A", end_color="92722A", fill_type="solid")
draft_cell_fill = PatternFill(start_color="E8F1FB", end_color="E8F1FB", fill_type="solid")
flag_fill = PatternFill(start_color="FDE2E2", end_color="FDE2E2", fill_type="solid")
thin = Side(style="thin", color="D9D9D9")
border = Border(left=thin, right=thin, top=thin, bottom=thin)
cell_font = Font(name="Arial", size=10)

for col_idx, name in enumerate(COLUMNS, start=1):
    c = ws.cell(row=1, column=col_idx, value=name)
    c.font = header_font
    if name in DRAFT_COLUMNS:
        c.fill = draft_header_fill
    elif name in VERIFIED_TRACKED:
        c.fill = verified_tracked_fill
    else:
        c.fill = header_fill
    c.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)
    c.border = border

draft_col_indexes = {i for i, name in enumerate(COLUMNS, start=1) if name in DRAFT_COLUMNS}
flags_col_index = COLUMNS.index("draftFlags") + 1
status_col_index = COLUMNS.index("draftStatus") + 1
status_col_letter = get_column_letter(status_col_index)

for r, row in enumerate(rows, start=2):
    for c_idx, key in enumerate(COLUMNS, start=1):
        value = cell_value(row, key)
        cell = ws.cell(row=r, column=c_idx, value=value if value != "" else None)
        cell.font = cell_font
        cell.border = border
        cell.alignment = Alignment(vertical="top", wrap_text=True)
        if c_idx == flags_col_index and value:
            cell.fill = flag_fill
        elif c_idx in draft_col_indexes and value:
            cell.fill = draft_cell_fill

# Restrict draftStatus to the three allowed values via a dropdown.
dv = DataValidation(type="list", formula1='"NEEDS REVIEW,APPROVED,REJECTED"', allow_blank=False)
ws.add_data_validation(dv)
dv.add(f"{status_col_letter}2:{status_col_letter}{len(rows) + 1}")

widths = {
    "slug": 34, "courseCode": 12, "courseName": 34, "category": 22, "subcategory": 22,
    "courseType": 14, "duration": 12, "curriculum": 50,
    "mode": 14, "level": 14, "overview": 34, "outcomes": 34, "eligibility": 34,
    "certification": 30, "fees": 18, "featured": 10, "status": 10, "notes": 26,
    "overviewDraft": 46, "outcomesDraft": 46, "curriculumDraft": 55,
    "draftStatus": 14, "draftFlags": 46,
}
for i, name in enumerate(COLUMNS, start=1):
    ws.column_dimensions[get_column_letter(i)].width = widths.get(name, 18)
ws.row_dimensions[1].height = 32

wb.save(OUT_DIR / "course-content-draft.xlsx")
print(f"Wrote {OUT_DIR / 'course-content-draft.xlsx'} ({len(rows)} rows)")
